### PR TITLE
THRIFT-4624: Fix refc binary leak

### PR DESCRIPTION
**Pull request by k32**
Client: erl

### CODE REVIEW

**Comment by jeking3** [comment]
I wonder if @walter-weinmann or @dhull could take a look at this perhaps?

File: lib/erl/src/thrift_compact_protocol.erl [hunk 1]
```diff
@@ -363,26 +365,32 @@ read(This0, double) ->
...
     {#t_compact{}, {ok, binary()} | {error, _Reason}}.
-read_data(This, 0) -> {This, {ok, <<>>}};
-read_data(This = #t_compact{transport = Trans}, Len) when is_integer(Len) andalso Len > 0 ->
-    {NewTransport, Result} = thrift_transport:read(Trans, Len),
+read_data(This, 0, _) -> {This, {ok, <<>>}};
+read_data(This, Len, Clone) when is_integer(Len) andalso Len > 0 ->
+    #t_compact{transport = Trans} = This,
+    {NewTransport, Result0} = thrift_transport:read(Trans, Len),
+    Result = case Result0 of
+                 {ok, Bin} when is_binary(Bin), Clone, Len > 64 ->
```
Comment (k32) [inline lib/erl/src/thrift_compact_protocol.erl]: The last condition should be removed. Cursory look at the BEAM code revealed that only brand new binaries can be allocated on heap, no copying is done for slices

Comment (jeking3) [reply lib/erl/src/thrift_compact_protocol.erl]: Are we missing a test case then? :)

Comment (k32) [reply lib/erl/src/thrift_compact_protocol.erl]: I'm not sure how one can reliably test this, as refc/heap binary dichotomy is hidden in the Erlang VM internals and there is no API to check which is which. The only (known to me) way to detect refc memory leak is through memory region monitoring, and it's beyond scope of unit/component tests. So regression testing is probably the best we can do.

Comment (jeking3) [reply lib/erl/src/thrift_compact_protocol.erl]: It sounded like you wanted to make one more change based on your original comment?

Comment (k32) [reply lib/erl/src/thrift_compact_protocol.erl]: Done

**Comment by jeking3** [comment]
All the tests pass however I'd like one independent code review before merging if possible.

**Comment by k32** [comment]
No hurry

**Comment by djnym** [comment]
The code seems reasonable, but I'm wondering of the need.  Do you actually see refc binary leakage?  I could maybe see that happen if you have long lived connections, from which you subtract a smallish portion of the binary from a much larger thrift object and pass it around, but I'm not totally certain.  Is there a case where you are actually seeing this leakage occuring?  @k32 Have you checked out https://github.com/openx/ox-thrift which @dhull wrote?  It uses the ranch acceptor pool instead of the standard server and is a complete rewrite of the message encoding/decoding.  It's about 5-10x faster at encoding/decoding, so it's an improvement.  The only way to make it faster would be to rewrite the compiler which @dhull has looked at but not done (the idea would be to generate a custom module for encoding/decoding instead of the generic structures which then get interpreted, this should lead to even more efficiency as you'd not have to go through an intermediate representation).  Anyway, maybe David will take a look at this, but I think it's probably fine to merge.  My only real comment would be that this appears to make the default copying the binaries, which could be a surprise if people aren't expecting it and are moving large thrift structures around.  So it might be worthwhile to document that somewhere, not sure where though.

**Comment by k32** [comment]
@djnym Thanks for the suggestion; on the first glance ox-thrift looks very promising. Regarding the defaults: it's just "safety first" approach. This can be changed.

**Comment by jeking3** [comment]
I'm curious to know the answer to:  "Is there a case where you are actually seeing this leakage occuring?"

**Comment by k32** [comment]
I found these errors via code review, while evaluating this library and doing some rapid prototyping. I'm not using Thrift in production. However, I would recommend merging at least part changing list_to_atom to list_to_existing_atom, because (if it wasn't obvious) it fixes a pretty bad DOS vulnerability, allowing attacker to take down the entire Erlang VM by filling up atom table.

**Comment by jeking3** [comment]
Are we all saying the changes in `lib/erl/src/thrift_processor.erl` are good but the others are unnecessary and may force additional copies of things unnecessarily?  Sorry I don't know erlang and no committer has stepped up to review.  Also `ox-thrift` looks interesting, but does not support as many transports or protocols as what exists here today and I assume the interface changed.  If it's that much better though (5-10x faster?) we should have a discussion about bringing that in, if the author desires it.  As for this one, I'm still not sure what to do because nobody made it clear.

**Comment by jeking3** [comment]
@djnym any thoughts about the comments regarding atom and DOS attacks?

**Comment by jeking3** [comment]
If you want to pare down the PR to just the changes in `lib/erl/src/thrift_processor.erl` and it passes the CI build, I can merge that.  I'm not sure the clone changes are needed per comments.

**Pull request closed without merging**